File: src/main/java/com/github/iamniklas/liocore/procedures/variants/FillStripProcedure.java
package com.github.iamniklas.liocore.procedures.variants;

import com.github.iamniklas.liocore.led.LEDDataBundle;
import com.github.iamniklas.liocore.led.LEDStripManager;
import com.github.iamniklas.liocore.led.colorspace.LIOColor;
import com.github.iamniklas.liocore.network.LEDUpdateModel;
import com.github.iamniklas.liocore.procedures.Procedure;
import com.github.iamniklas.liocore.procedures.models.Direction;
import org.checkerframework.checker.units.qual.C;

public class FillStripProcedure extends Procedure {

    private LEDDataBundle bundle;

    private LIOColor fillColor = new LIOColor(0, 0, 0);
    private int speed = 1;
    private Direction direction = Direction.Left;

    public FillStripProcedure(LEDUpdateModel _ledUpdateModel) {
        super(_ledUpdateModel);
        bundle = _ledUpdateModel.bundle;

        fillColor = LIOColor.fromRGB(bundle.colorPrimary);
        speed = Math.round(bundle.speed);
        direction = bundle.direction;
        steps = LEDStripManager.ledCount;

        if(direction == Direction.Center || direction == Direction.CenterInvert) {
            steps /= 2;
        }
    }

    @Override
    public void start() {
        super.start();
    }

    @Override
    public void update() {
        int i = 0;
        switch (direction) {
            case Left:
                while (i < speed) {
                    strip.setPixel(step + i, fillColor);
                    i++;
                }
                break;
            case Right:
                while (i < speed) {
                    strip.setPixel(LEDStripManager.ledCount - 1 - step - i, fillColor);
                    i++;
                }
                break;
            case Center:
                while (i < speed) {
                    strip.setPixel(LEDStripManager.ledCount / 2 - 1 + step + i, fillColor);
                    strip.setPixel(LEDStripManager.ledCount / 2 - 1 - step - i, fillColor);
                    i++;
                }
                break;
            case CenterInvert:
                while (i < speed) {
                    strip.setPixel(step + i, fillColor);
                    strip.setPixel(LEDStripManager.ledCount - 1 - step - i, fillColor);
                    i++;
                }
                break;
        }

        step += speed;
        if(step >= steps) {
            finishProcedure();
        }
    }

    @Override
    public void updateLEDUpdateModel(LEDUpdateModel _ledUpdateModel) {

    }
}

File: src/main/java/com/github/iamniklas/liocorekotlin/procedures/FillStripProcedure.kt
package com.github.iamniklas.liocorekotlin.procedures

import com.github.iamniklas.liocorekotlin.led.LEDDataBundle
import com.github.iamniklas.liocorekotlin.led.LEDStripManager
import procedures.models.Direction
import java.awt.Color

//Animation to fill the strip from a given direction (left, right, center, bounds) with a given color 
class FillStripProcedure(_bundle: com.github.iamniklas.liocorekotlin.led.LEDDataBundle) : com.github.iamniklas.liocorekotlin.procedures.Procedure(_bundle) {
    private var mFillColor = Color.BLACK
    var mSpeed = 1
    var mDirection: Direction? = Direction.Left
    override fun start() {}
    public override fun update() {
        when (mDirection) {
            Direction.Center -> {
                var i = 0
                while (i < mSpeed) {
                    mStrip!!.setPixel(com.github.iamniklas.liocorekotlin.led.LEDStripManager.LED_COUNT / 2 - 1 + mStep + i, mFillColor)
                    mStrip!!.setPixel(com.github.iamniklas.liocorekotlin.led.LEDStripManager.LED_COUNT / 2 - 1 - mStep - i, mFillColor)
                    i++
                }
            }
            Direction.CenterInvert -> {
                var i = 0
                while (i < mSpeed) {
                    mStrip!!.setPixel(mStep + i, mFillColor)
                    mStrip!!.setPixel(com.github.iamniklas.liocorekotlin.led.LEDStripManager.LED_COUNT - 1 - mStep - i, mFillColor)
                    i++
                }
            }
            Direction.Left -> {
                var i = 0
                while (i < mSpeed) {
                    mStrip!!.setPixel(mStep + i, mFillColor)
                    i++
                }
            }
            Direction.Right -> {
                var i = 0
                while (i < mSpeed) {
                    mStrip!!.setPixel(com.github.iamniklas.liocorekotlin.led.LEDStripManager.LED_COUNT - 1 - mStep - i, mFillColor)
                    i++
                }
            }
            else -> {
            }
        }
        mStep += mSpeed
        if (mStep >= mSteps) {
            if (!mIsSubProcedure) {
                finishProcedure()
            }
        }
    }

    init {
        mFillColor = _bundle.colorPrimary!!.toSystemColor()
        mSpeed = Math.round(_bundle.speed!!)
        mDirection = _bundle.direction
        mSteps = com.github.iamniklas.liocorekotlin.led.LEDStripManager.LED_COUNT
        if (mDirection == Direction.Center || mDirection == Direction.CenterInvert) {
            mSteps /= 2
        }
    }
}